File: Card Transition Demo/Animation/KAKAnimation.m
//
//  KAKAnimation.m
//  Card Transition Demo
//
//  Created by Mingfei Huang on 11/4/18.
//  Copyright © 2018 Mingfei Huang. All rights reserved.
//

#import "KAKAnimation.h"

#import <YYKit/YYKit.h>

@interface KAKAnimation ()

// frame of `self.triggerView` in coordinate of container view
@property (assign, nonatomic) CGRect triggerViewFrame;

@property (strong, nonatomic) UIView *triggerViewParentView;

@end

@implementation KAKAnimation

+ (instancetype)instanceWithTriggerView:(UIView *)triggerView
               presentingViewController:(UIViewController<KAKAnimationPresentingViewController> *)presentingViewController
                presentedViewController:(UIViewController<KAKAnimationPresentedViewController> *)presentedViewController
{
    KAKAnimation *inst = self.new;
    if (!inst) {
        return nil;
    }
    inst->_triggerView = triggerView;
    inst->_presentingViewController = presentingViewController;
    inst->_presentedViewController = presentedViewController;
    return inst;
}

- (void)dismissInteractively:(BOOL)interactively
{
    _dismissInteractively = interactively;
    [self.presentedViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)presentAnimate:(id<UIViewControllerContextTransitioning>)context
         containerView:(UIView *)containerView
              fromView:(UIView *)fromView
                toView:(UIView *)toView
    fromViewController:(UIViewController *)fromViewController
      toViewController:(UIViewController *)toViewController
{
    [containerView addSubview:fromView];
    [containerView addSubview:toView];
    
    // store initial frame and parent view for restoration
    self.triggerViewFrame = [self.triggerView.superview convertRect:self.triggerView.frame toView:containerView];
    self.triggerViewParentView = self.triggerView.superview;
    
    // prepare before anaimation
    toView.frame = self.triggerViewFrame;
    [self.presentedViewController preparePresentAnimation:self.triggerView];
    
    self.triggerView.frame = [containerView convertRect:self.triggerViewFrame toView:self.triggerView.superview];
    
    // do present animation
    [UIView animateWithDuration:[self transitionDuration:context] delay:0 usingSpringWithDamping:0.7 initialSpringVelocity:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        toView.frame = [context finalFrameForViewController:toViewController];
        [self.presentedViewController performPresentAnimation:self.triggerView];
    } completion:^(BOOL finished) {
        [fromView removeFromSuperview];
        [context completeTransition:YES];
    }];
}

- (void)dismissAnimate:(id<UIViewControllerContextTransitioning>)context
         containerView:(UIView *)containerView
              fromView:(UIView *)fromView
                toView:(UIView *)toView
    fromViewController:(UIViewController *)fromViewController
      toViewController:(UIViewController *)toViewController
{
    [containerView addSubview:toView];
    [containerView addSubview:fromView];
    fromView.frame = [context initialFrameForViewController:fromViewController];
    
    [UIView animateKeyframesWithDuration:[self transitionDuration:context] delay:0 options:0 animations:^{
        // 1. shrink detail view
        [UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.5 animations:^{
            CGPoint oldCenter = fromView.center;
            fromView.width *= 0.7;
            fromView.height *= 0.7;
            fromView.center = oldCenter;
            [fromView layoutIfNeeded];
        }];
        
        // 2. dismiss detail view, shrink trigger view back to its original position
        [UIView addKeyframeWithRelativeStartTime:0.5 relativeDuration:0.5 animations:^{
            fromView.frame = self.triggerViewFrame;
            self.triggerView.frame = [containerView convertRect:self.triggerViewFrame toView:self.triggerView.superview];
        }];
    } completion:^(BOOL finished) {
        if (!context.transitionWasCancelled) {
            self.triggerView.frame = [containerView convertRect:self.triggerViewFrame toView:self.triggerViewParentView];
            [self.triggerViewParentView addSubview:self.triggerView];
            [fromView removeFromSuperview];
        } else {
            // only restore first stage of animation
            // second stage of animation is ireversable
            // it should be controlled by interaction control logic
            [toView removeFromSuperview];
        }
        [context completeTransition:!context.transitionWasCancelled];
    }];
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext
{
    UIViewController *fromViewController = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    if (fromViewController == self.presentingViewController) {
        [self presentAnimate:transitionContext
               containerView:transitionContext.containerView
                    fromView:[transitionContext viewForKey:UITransitionContextFromViewKey]
                      toView:[transitionContext viewForKey:UITransitionContextToViewKey]
          fromViewController:[transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey]
            toViewController:[transitionContext viewControllerForKey:UITransitionContextToViewControllerKey]];
    } else {
        [self dismissAnimate:transitionContext
               containerView:transitionContext.containerView
                    fromView:[transitionContext viewForKey:UITransitionContextFromViewKey]
                      toView:[transitionContext viewForKey:UITransitionContextToViewKey]
          fromViewController:[transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey]
            toViewController:[transitionContext viewControllerForKey:UITransitionContextToViewControllerKey]];
    }
}

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext
{
    return .5;
}

#pragma mark - UIViewControllerTransitioningDelegate

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source
{
    return self;
}

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed
{
    return self;
}

- (id<UIViewControllerInteractiveTransitioning>)interactionControllerForDismissal:(id<UIViewControllerAnimatedTransitioning>)animator
{
    if (self.dismissInteractively) {
        return self;
    } else {
        return nil;
    }
}

- (void)cancelInteractiveTransition
{
    [super cancelInteractiveTransition];
    
    for (UIView *view in self.presentedViewController.view.superview.subviews) {
        if (view != self.presentedViewController.view) {
            [view removeFromSuperview];
        }
    }
}

@end

File: Card Transition Demo/Animation/KAKAnimation.swift
//
//  KAKAnimation.swift
//  Card Transition Demo
//
//  Created by Mingfei Huang on 11/4/18.
//  Copyright © 2018 Mingfei Huang. All rights reserved.
//

import YYKit

protocol KAKAnimationPresentingViewController: NSObjectProtocol {
    /**
     *  a callback to presenting view controller by end of dismissal animation
     *  called after trigger view has been passed back to its original view hierarchy
     */
    func completeDismissAnimation(_ triggerView: UIView)
}

protocol KAKAnimationPresentedViewController: NSObjectProtocol {
    /**
     *  a callback to presented view controller by begining of present animation
     *  presented view controller should use this callback to install trigger view into its view hierarchy
     */
    func preparePresentAnimation(_ triggerView: UIView)
    
    /**
     *  a callback to presented view controller by end of present animation
     *  presented view controller should apply finalized style to trigger view
     */
    func performPresentAnimation(_ triggerView: UIView)
}


class KAKAnimation: UIPercentDrivenInteractiveTransition {
    var triggerViewFrame: CGRect?
    var triggerViewParentView: UIView?
    
    var triggerView: UIView
    var presentingViewController: (KAKAnimationPresentingViewController & UIViewController)
    var presentedViewController: (KAKAnimationPresentedViewController & UIViewController)
    
    private var dismissInteractively = false
    
    init(
        triggerView: UIView,
        presentingViewController: (KAKAnimationPresentingViewController & UIViewController),
        presentedViewController: (KAKAnimationPresentedViewController & UIViewController)
        ) {
        self.triggerView = triggerView
        self.presentingViewController = presentingViewController
        self.presentedViewController = presentedViewController
        super.init()
    }
    
    func dismiss(interactively: Bool) {
        self.dismissInteractively = interactively
        self.presentedViewController.dismiss(animated: true, completion: nil)
    }
}

extension KAKAnimation: UIViewControllerAnimatedTransitioning {
    func presentAnimateWith(
        context: UIViewControllerContextTransitioning,
        containerView: UIView,
        fromView: UIView,
        toView: UIView,
        fromViewController: UIViewController,
        toViewController: UIViewController
        )
    {
        containerView.addSubview(fromView)
        containerView.addSubview(toView)
        
        // store initial frame and parent view for restoration
        self.triggerViewFrame = self.triggerView.superview?.convert(self.triggerView.frame, to: containerView)
        self.triggerViewParentView = self.triggerView.superview
        
        // prepare before anaimation
        toView.frame = self.triggerViewFrame!
        self.presentedViewController.preparePresentAnimation(self.triggerView)
        
        self.triggerView.frame = containerView.convert(self.triggerViewFrame!, to: self.triggerView.superview)
        
        // do present animation
        UIView.animate(withDuration: self.transitionDuration(using: context), animations: {
            toView.frame = context.finalFrame(for: toViewController)
            self.presentedViewController.performPresentAnimation(self.triggerView)
        }) { (_) in
            fromView.removeFromSuperview()
            context.completeTransition(true)
        }
    }
    
    func dismissAnimateWith(
        context: UIViewControllerContextTransitioning,
        containerView: UIView,
        fromView: UIView,
        toView: UIView,
        fromViewController: UIViewController,
        toViewController: UIViewController
        )
    {
        containerView.addSubview(toView)
        containerView.addSubview(fromView)
        fromView.frame = context.initialFrame(for: fromViewController)
        let triggerViewFrame = self.triggerViewFrame ?? CGRect.zero
        
        UIView.animateKeyframes(withDuration: self.transitionDuration(using: context), delay: 0, options: [], animations: {
            
            // 1. shrink detail view
            UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 0.5, animations: {
                let oldCenter = fromView.center
                fromView.width *= 0.7
                fromView.height *= 0.7
                fromView.center = oldCenter
                fromView.layoutIfNeeded()
            })
            
            // 2. dismiss detail view, shrink trigger view back to its original position
            UIView.addKeyframe(withRelativeStartTime: 0.5, relativeDuration: 0.5, animations: {
                fromView.frame = self.triggerViewFrame ?? CGRect.zero
                self.triggerView.frame = containerView.convert(triggerViewFrame, to: self.triggerView.superview)
            })
        }) { (completed) in
            if (context.transitionWasCancelled) {
                toView.removeFromSuperview()
            } else {
                self.triggerView.frame = containerView.convert(triggerViewFrame, to: self.triggerViewParentView)
                self.triggerViewParentView?.addSubview(self.triggerView)
                fromView.removeFromSuperview()
            }
            context.completeTransition(!context.transitionWasCancelled)
        }
    }
    
    func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        let fromViewController = transitionContext.viewController(forKey: UITransitionContextViewControllerKey.from)
        if (fromViewController == self.presentingViewController) {
            self.presentAnimateWith(context: transitionContext,
                                    containerView: transitionContext.containerView,
                                    fromView: transitionContext.view(forKey: UITransitionContextViewKey.from)!,
                                    toView: transitionContext.view(forKey: UITransitionContextViewKey.to)!,
                                    fromViewController: transitionContext.viewController(forKey: UITransitionContextViewControllerKey.from)!,
                                    toViewController: transitionContext.viewController(forKey: UITransitionContextViewControllerKey.to)!)
        } else {
            self.dismissAnimateWith(context: transitionContext,
                                    containerView: transitionContext.containerView,
                                    fromView: transitionContext.view(forKey: UITransitionContextViewKey.from)!,
                                    toView: transitionContext.view(forKey: UITransitionContextViewKey.to)!,
                                    fromViewController: transitionContext.viewController(forKey: UITransitionContextViewControllerKey.from)!,
                                    toViewController: transitionContext.viewController(forKey: UITransitionContextViewControllerKey.to)!)
        }
    }
    
    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        return 0.5
    }
}

extension KAKAnimation: UIViewControllerTransitioningDelegate {
    func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return self
    }
    
    func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return self
    }
    
    func interactionControllerForDismissal(using animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        if (self.dismissInteractively) {
            return self
        } else {
            return nil
        }
    }
}
